controller/engine_image_controller.go

```diff
@@ -378,9 +380,152 @@ func (ic *EngineImageController) syncEngineImage(key string) (err error) {
+
+	// Upgrade engine image for volume
+	for _, v := range upgradingCandidates {
+		oldEngineImage := v.Spec.EngineImage
+		v.Spec.EngineImage = defaultEngineImage
+		v, err = ic.ds.UpdateVolume(v)
+		if err != nil {
+			return err
+		}
+		ic.logger.Infof("automatically upgrades volume's engine image from %v to the default engine image %v for volume %v", oldEngineImage, defaultEngineImage, v.Name)
+	}
+
+	return nil
+}
+
+func limitAutomaticEngineUpgradePerNode(volumes map[string]*longhorn.Volume, upgradingCandidates []*longhorn.Volume) []*longhorn.Volume {
+	finalCandidates := []*longhorn.Volume{}
+	candidatePerNode := map[string]int{}
+	// Count number of volumes per node that are upgrading engine
+	for _, v := range volumes {
+		if v.Spec.EngineImage != v.Status.CurrentImage {
+			nodeName := v.Status.OwnerID
+			candidatePerNode[nodeName] = candidatePerNode[nodeName] + 1
+		}
+	}
+	// Only allow concurrentAutomaticEngineUpgradePerNodeLimit engines to be automatically upgraded at a time
+	for _, v := range upgradingCandidates {
+		node := v.Status.OwnerID
+		if candidatePerNode[node] < concurrentAutomaticEngineUpgradePerNodeLimit {
+			candidatePerNode[node] = candidatePerNode[node] + 1
+			finalCandidates = append(finalCandidates, v)
+		}
+	}
+	return finalCandidates
+}
+
+// getVolumesForEngineImageUpgrading returns a list of volumes that are qualified for engine image upgrading
+// A volume is qualified for engine image upgrading if it meets one of the following case:
+// Case 1:
+//   1. Volume is in detached state
+// Case 2:
+//   1. Volume is not in engine upgrading process
+//   2. Volume is in attached state and it is able to do live upgrade
+func (ic *EngineImageController) getVolumesForEngineImageUpgrading(volumes map[string]*longhorn.Volume, newEngineImageResource *longhorn.EngineImage) ([]*longhorn.Volume, error) {
+	candiates := []*longhorn.Volume{}
+
+	for _, v := range volumes {
+		// Already attempted to upgrade volume's engine image
+		if v.Spec.EngineImage == newEngineImageResource.Spec.Image {
+			continue
+		}
+		// Volume is in upgrading process. Need to wait for the upgrade to finish.
+		if v.Spec.EngineImage != v.Status.CurrentImage {
+			continue
+		}
+		if !ic.canDoOfflineEngineImageUpgrade(v, newEngineImageResource) && !ic.canDoLiveEngineImageUpgrade(v, newEngineImageResource) {
+			continue
+		}
+		candiates = append(candiates, v)
+	}
+
+	return candiates, nil
+}
+
+func (ic *EngineImageController) canDoOfflineEngineImageUpgrade(v *longhorn.Volume, newEngineImageResource *longhorn.EngineImage) bool {
+	return v.Status.State == types.VolumeStateDetached && newEngineImageResource.Status.State == types.EngineImageStateReady
+}
+
+// canDoLiveEngineImageUpgrade check if it is possible to do live engine upgrade for a volume
+// A volume can do live engine upgrade when:
+//   1. Volume is attached AND
+//   2. Volume's robustness is either healthy or degraded AND
```

Comment:
Cannot do live upgrade for degraded volumes:
https://github.com/longhorn/longhorn-manager/blob/2609cf4f61c00783d621f985567b83bc4c6e052a/controller/volume_controller.go#L1577

Reply:
Thank you. Fixed the degraded volumes upgrading.